Sources/SourceKitLSP/SourceKitLSPServer.swift

```diff
@@ -1238,6 +1257,38 @@ extension SourceKitLSPServer {
     }
   }
 
+  /// Cancel all in-progress text document requests for the given document.
+  ///
+  /// As a user makes an edit to a file, these requests are most likely no longer relevant. It also makes sure that a
+  /// long-running sourcekitd request can't block the entire language server if the client does not cancel all requests.
+  /// For example, consider the following sequence of requests:
+  ///  - `textDocument/semanticTokens/full` for document A
+  ///  - `textDocument/didChange` for document A
+  ///  - `textDocument/formatting` for document A
+  ///
+  /// If the editor is not cancelling the semantic tokens request on edit (like VS Code does), then the `didChange`
+  /// notification is blocked on the semantic tokens request finishing. Hence, we also can't run the
+  /// `textDocument/formatting` request. Cancelling the semantic tokens on the edit fixes the issue.
+  ///
+  /// This method is a no-op if `cancelTextDocumentRequestsOnEditAndClose` is disabled.
+  private nonisolated func cancelTextDocumentRequests(for uri: DocumentURI) {
+    // Since the request is very cheap to execute and stops other requests
+    // from performing more work, we execute it with a high priority.
+    cancellationMessageHandlingQueue.async(priority: .high) {
+      await self.cancelTextDocumentRequestsImpl(for: uri)
+    }
+  }
+
+  private func cancelTextDocumentRequestsImpl(for uri: DocumentURI) {
+    guard self.options.cancelTextDocumentRequestsOnEditAndCloseOrDefault else {
+      return
+    }
```

Comment:
Could just check in the outer method before queuing anything?

Reply:
No, because `options` is isolated to the `SourceKitLSPServer` instance and the outer `cancelTextDocumentRequests` method is nonisolated.